handlers/api/callback.go

```diff
@@ -131,3 +102,74 @@ func CallbackGet(c *fiber.Ctx) error {
 
 	return c.Redirect("/account", fiber.StatusSeeOther)
 }
+
+func flow(o oauthlogin.OAuthResponse) (*models.User, error) {
+	var eu models.ExternalUser
+
+	// Check if external user exists.
+	err := database.Conn.Debug().
+		Model(eu).Preload("User").
+		First(&eu, "provider = ? AND external_id = ?", o.Provider, o.ExternalID).Error
+
+	switch {
+	case err == gorm.ErrRecordNotFound:
+		// Check if user exists.
+		err := database.Conn.Debug().
+			First(&eu.User, "username = ? COLLATE NOCASE AND o_auth_provider = ?", o.Username, o.Provider).Error
+		if err != nil {
+			eu = models.ExternalUser{
+				ExternalID:  o.ExternalID,
+				Provider:    string(o.Provider),
+				Email:       o.Email,
+				Username:    o.Username,
+				ExternalURL: o.ProfileURL(),
+				AccessToken: o.AccessToken,
+				RawData:     o.RawData,
+				User: models.User{
+					Email:         strings.ToLower(o.Email),
+					Username:      strings.ToLower(o.Username),
+					OAuthProvider: string(o.Provider),
+				},
+			}
+
+			var count int64
+			err := database.Conn.Debug().
+				Model(eu.User).
+				Where("username = ? OR email = ?", eu.Username, eu.Email).
+				Count(&count).
+				Error
+			if err != nil || count > 0 {
+				return nil, fmt.Errorf("user already exists")
+			}
+
+			if err := database.Conn.Debug().Create(&eu).Error; err != nil {
```

Comment:
Shouldn't we create a user as well then? As from my understanding this means someone tries to register via OAuth.

Reply:
That's correct. GORM handles it automatically for us.

Reply:
That's correct. GORM handles it automatically for us.